File: search/searchA2DMatrixII.py
# https://leetcode.com/problems/search-a-2d-matrix-ii/

from typing import List
class Solution:
    
    # Runtime: 189 ms, faster than 84.28% of Python3 online submissions for Search a 2D Matrix II.
    # Memory Usage: 20.5 MB, less than 13.29% of Python3 online submissions for Search a 2D Matrix II.
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        
        row , col = 0, len(matrix[0])-1
        while row<len(matrix) and  col>=0:
        
            pointing = matrix[row][col]
            if pointing == target : return True
            if pointing > target: col-=1
            else : row+=1
                
        return False

File: search/searchA2DMatrixII.java
package search;

public class searchA2DMatrixII {
    public static void main(String[] args) {
    SolutionSA2DMII solution = new SolutionSA2DMII();
    System.out.println(solution.searchMatrix(new int[][]{{1,4,7,11,15},{2,5,8,12,19},{3,6,9,16,22},{10,13,14,17,24},{18,21,23,26,30}},5));
    }
}

class SolutionSA2DMII {
    public boolean searchMatrix(int[][] matrix, int target) {
        
        int row = 0;
        int col = matrix[0].length-1;
        
        while(row<matrix.length && col>=0){
            int pointing = matrix[row][col];
            if (pointing == target) return true;
            else if ( pointing> target) col--;
            else row++;
        }
        
        return false;
        
    }
}